lib/storage/data/external/GCP/GcpService.js

```diff
@@ -1,141 +1,908 @@
-const AWS = require('aws-sdk');
-const Service = AWS.Service;
-
+// GcpClient.js: GCP-compatible S3 client using AWS SDK v3
 const errorInstances = require('../../../../errors').errorInstances;
-const GcpApis = require('./GcpApis');
-const GcpServiceSetup = require('./GcpServiceSetup');
-const GcpManagedUpload = require('./GcpManagedUpload');
+const constructStringToSignV2 = require('../../../../auth/v2/constructStringToSign').default;
+const crypto = require('crypto');
+const getPutTagsMetadata = require('./GcpUtils').getPutTagsMetadata;
+const processTagSet = require('./GcpUtils').processTagSet;
+const GcpManagedUpload = require('../GCP/GcpManagedUpload');
+const async = require('async');
+const createMpuKey = require('./GcpUtils').createMpuKey;
+const getPartNumber = require('./GcpUtils').getPartNumber;
+const MpuHelper = require('./GcpApis/mpuHelper');
+const { v4: uuid } = require('uuid');
+const stripTags = require('./GcpUtils').stripTags;
+const retrieveTags = require('./GcpUtils').retrieveTags;
+const { S3Client,
+    PutObjectCommand,
+    GetObjectCommand,
+    DeleteObjectCommand,
+    CopyObjectCommand,
+    HeadObjectCommand,
+    HeadBucketCommand,
+    ListObjectsCommand,
+    CreateMultipartUploadCommand,
+    UploadPartCommand,
+    CompleteMultipartUploadCommand,
+    AbortMultipartUploadCommand,
+    PutObjectTaggingCommand,
+    GetObjectTaggingCommand,
+    DeleteObjectTaggingCommand,
+    ListPartsCommand,
+    GetBucketVersioningCommand,
+    PutBucketVersioningCommand,
+} = require('@aws-sdk/client-s3');
+
+/**
+ * GcpClient: S3-compatible client for Google Cloud Storage using AWS SDK v3.
+ *
+ * Configuration options:
+ *   - s3Params: AWS SDK v3 S3Client configuration (endpoint, credentials, etc.)
+ *   - bucketName: Main GCP bucket name
+ *   - mpuBucket: Bucket for multipart uploads (optional, defaults to bucketName)
+ *   - dataStoreName: Logical name for the backend
+ *   - type: Should be 'gcp'
+ *
+ * GCP-specific behaviors:
+ *   - Object tags are stored as metadata with the prefix 'scal-s3b-tag-'.
+ *   - MPU (multipart upload) is emulated using GCP's compose/copy APIs.
+ *   - All S3-compatible methods are available, with overrides for GCP quirks.
+ */
+class GcpClient extends S3Client {
+    /**
+     * @param {object} config - Configuration object (see above)
+     */
+    constructor(config) {
+        super(config.s3Params);
+        this._config = config;
+        this._maxConcurrent = 10;
+
+        this.middlewareStack.remove('httpSigningMiddleware');
+
+        // Add middleware to normalize headers before they reach signing middleware
+        // AWS SDK v3 can pass header values as arrays or non-string types,
+        // but signing libraries expect strings. This middleware normalizes all
+        // header values to strings before signing.
+        this.middlewareStack.add(
+            (next) => async (args) => {
```

Comment:
```suggestion
            next => async args => {
```

(many places below as well)